packages/roosterjs-editor-plugins/lib/plugins/ContentEdit/features/markdownFeatures.ts

```diff
@@ -0,0 +1,127 @@
+import { ContentEditFeature, Editor, Keys } from 'roosterjs-editor-core';
+import { createRange } from 'roosterjs-editor-dom';
+
+const ZERO_WIDTH_SPACE = '\u200B';
+
+function generateBasicMarkdownFeature(
+    key: Keys,
+    triggerCharacter: string,
+    elementTag: string,
+    useShiftKey: boolean
+): ContentEditFeature {
+    return {
+        keys: [key],
+        shouldHandleEvent: (event, editor) =>
+            event.rawEvent.shiftKey === useShiftKey &&
+            !!getRangeForMarkdownOperation(event, editor, triggerCharacter),
+        handleEvent: (event, editor) => {
+            event.rawEvent.preventDefault();
+            handleMarkdownEvent(event, editor, triggerCharacter, elementTag);
+        },
+    };
+}
+
+function getRangeForMarkdownOperation(
+    event: PluginKeyboardEvent,
+    editor: Editor,
+    triggerCharacter: string
+): Range {
+    const searcher = editor.getContentSearcherOfCursor();
+
+    let startPosition: NodePosition;
+    let endPosition: NodePosition;
+    searcher.forEachTextInlineElement(textInlineElement => {
+        if (endPosition && startPosition) {
+            return true;
+        }
+        const inlineTextContent = textInlineElement.getTextContent();
+        if (inlineTextContent[inlineTextContent.length - 1].trim().length == 0) {
+            return false;
+        }
+        endPosition = textInlineElement.getStartPosition().move(inlineTextContent.length);
+        if (inlineTextContent[0] == triggerCharacter) {
+            startPosition = textInlineElement.getStartPosition();
+        } else {
+            let contentIndex = inlineTextContent.length - 1;
+            for (; contentIndex > 0; contentIndex--) {
+                if (startPosition) {
+                    return true;
+                }
+                if (
+                    inlineTextContent[contentIndex] == triggerCharacter &&
+                    inlineTextContent[contentIndex - 1].trim().length == 0
+                ) {
+                    startPosition = textInlineElement.getStartPosition().move(contentIndex);
+                    return true;
+                }
+            }
+        }
+    });
+    return !!startPosition && !!endPosition && createRange(startPosition, endPosition);
+}
+
+function handleMarkdownEvent(
+    event: PluginKeyboardEvent,
+    editor: Editor,
+    triggerCharacter: string,
+    elementTag: string
+) {
+    editor.performAutoComplete(() => {
+        const range = getRangeForMarkdownOperation(event, editor, triggerCharacter);
+        if (!!range) {
+            const elementToWrap = editor.getDocument().createElement(elementTag);
+            elementToWrap.appendChild(range.extractContents());
+            elementToWrap.innerText = elementToWrap.innerText.slice(
+                1,
+                elementToWrap.innerText.length
+            );
+            const nonPrintedSpaceTextNode = editor.getDocument().createTextNode(ZERO_WIDTH_SPACE);
```

Comment:
why do we need this?

Reply:
We need to put the cursor outside of the created wrapping element. By adding a ZWS we clearly separate the newly created bold/italic/strikethrough block and are able to put the cursor somewhere where they can keep typing normally.

Reply:
Is there a cleaner way to do this, in your opinion? I tried fiddling with the range objects but wasn't able to get it to work.

Reply:
We can just select after the new node, then call toggleBold/Italic/... to force turn off current state

Reply:
Doesn't work for `code` as a feature.

Reply:
Ok. We can go with ZWS for now, since font and color formats are also using it.
I'll think about a way to apply format without using ZWS later.